tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ConstSimplifierBase.scala

```diff
@@ -29,49 +29,170 @@ abstract class ConstSimplifierBase {
+
+    // 0 + x = x
+    case OperEx(TlaArithOper.plus, ValEx(TlaInt(left)), rightEx) =>
       if (left == 0) {
-        ValEx(TlaInt(right))(intTag)
-      } else if (right == 0) {
-        ValEx(TlaInt(left))(intTag)
+        rightEx
+      } else {
+        OperEx(TlaArithOper.plus, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
+
+    // x + 0 = x
+    case OperEx(TlaArithOper.plus, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        leftEx
       } else {
-        ValEx(TlaInt(left + right))(intTag)
+        OperEx(TlaArithOper.plus, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
       }
 
     case OperEx(TlaArithOper.minus, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
       ValEx(TlaInt(left - right))(intTag)
 
-    case ex @ OperEx(TlaArithOper.minus, NameEx(left), NameEx(right)) =>
-      if (left == right) ValEx(TlaInt(0))(intTag) else ex // this actually happens
+    // 0 - x = -x
+    case OperEx(TlaArithOper.minus, ValEx(TlaInt(left)), rightEx) =>
+      if (left == 0) {
+        OperEx(TlaArithOper.uminus, rightEx)(intTag)
+      } else {
+        OperEx(TlaArithOper.minus, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
+
+    // x - 0 = x
+    case OperEx(TlaArithOper.minus, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.minus, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    // x - x = 0
+    case ex @ OperEx(TlaArithOper.minus, leftEx, rightEx) =>
+      if (leftEx == rightEx) ValEx(TlaInt(0))(intTag) else ex // this actually happens
 
     case OperEx(TlaArithOper.mult, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
-      if (left == 0 || right == 0) {
-        ValEx(TlaInt(BigInt(0)))(intTag)
+      ValEx(TlaInt(left * right))(intTag)
+
+    // 0 * x = 0
+    // 1 * x = x
+    case OperEx(TlaArithOper.mult, ValEx(TlaInt(left)), rightEx) =>
+      if (left == 0) {
+        ValEx(TlaInt(0))(intTag)
       } else if (left == 1) {
-        ValEx(TlaInt(right))(intTag)
+        rightEx
+      } else {
+        OperEx(TlaArithOper.mult, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
+
+    // x * 0 = 0
+    // x * 1 = x
+    case OperEx(TlaArithOper.mult, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        ValEx(TlaInt(0))(intTag)
+      } else if (right == 1) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.mult, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    case ex @ OperEx(TlaArithOper.div, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Division by zero at ${ex.toString}")
+      } else {
+        ValEx(TlaInt(left / right))(intTag)
+      }
+
+    // 0 / x = 0
+    case OperEx(TlaArithOper.div, ValEx(TlaInt(left)), rightEx) =>
+      if (left == 0) {
+        ValEx(TlaInt(0))(intTag)
+      } else {
+        OperEx(TlaArithOper.div, ValEx(TlaInt(left))(intTag), rightEx)(intTag)
+      }
```

Comment:
The best way to deal with undefined behavior is reporting it instead of trying to compute anything, but that's not always possible. In my understanding, the changes on this PR helps a bit with reporting in some cases and does not actually change the result in other cases. It just changes what is computed when behavior is undefined, but there's no better or worse in that, right? Here's some examples:
 
Expressions like `x = 0 /\ e = 0 / x`
Before: `e` can be anything
Now: `e = 0`
None of them throws an error which would be the best thing to do on undefined behavior. Not better not worse.

Expressions like `x = 1 /\ e = 0 / (x - x)``
Before: `e` can be anything
Now: Throws an error
Better than the current implementation.

PS: This assumes that the SMT solver is actually ignoring (evaluating as `true`) any expressions with undefined behavior, which seems to be the case for the examples I tested.

Regarding your example, I've tested on both this branch and `unstable` and the behavior is the same. Probably because the solver is just setting the hole `(0 = 0 / x)` expression to true when the operation is undefined.

I believe we should document better what we do with undefined behavior, which ones we can detect and maybe even explore adding constraints as you suggested earlier this week. I still have to open an issue for that as we agreed, and my suggestion would be to go further in this discussion when we address the issue. That is, if you agree that this PR is not making anything worse.

Reply:
Again, I support all of the simplifications introduced in the PR in the dynamically undefined cases, on the condition that we have a document (or at least comments) explicitly mentioning that the cases were considered and a design decision was made to e.g. treat `0^0` as 1 for consistency.
This PR is long enough as it is, but it might make sense to open an issue for adding these decisions to the manual.